### DISCUSSION TITLE
Filters RFC

### DISCUSSION THREAD

**Discussion by soyuka** [api-platform/api-platform#1724] [General]
We're looking for popular opinions about filters RFC. Here are two examples of well-done filter composition:

- Loopback: https://loopback.io/doc/en/lb3/Where-filter.html
- Postgrest: https://postgrest.org/en/v7.0.0/api.html#horizontal-filtering-rows

[![](https://api.gh-polls.com/poll/01ESVYGN4ZW5GYXBZ1QC12MD6Y/Loopback)](https://api.gh-polls.com/poll/01ESVYGN4ZW5GYXBZ1QC12MD6Y/Loopback/vote)
[![](https://api.gh-polls.com/poll/01ESVYGN4ZW5GYXBZ1QC12MD6Y/Postgrest)](https://api.gh-polls.com/poll/01ESVYGN4ZW5GYXBZ1QC12MD6Y/Postgrest/vote)
[![](https://api.gh-polls.com/poll/01ESVYGN4ZW5GYXBZ1QC12MD6Y/Other)](https://api.gh-polls.com/poll/01ESVYGN4ZW5GYXBZ1QC12MD6Y/Other/vote)

Do you know other similar APIs that could inspire us? Please share!

Linked issues:

- https://github.com/api-platform/api-platform/issues/639
- https://github.com/api-platform/core/issues/2400

**Comment by soyuka** [comment]
About `other` what motivated the choice?

**Comment by walva** [comment]
I developed another solution for my current project. I needed to provide a **single input** to my users to search in several fields. They will be able to search for example in firstname, lastname, username and email using only one query parameter.

The annotation look like this:
```
#[ApiFilter(MultiSearchFilter::class, properties: [
    'username' => 'partial',
    'firstName' => 'istart',
    'lastName' => 'istart',
    'email' => 'partial',
    ],
    arguments: [
        'separator' => ',',
        'explodeEnabled' => true,
        'overrideExplodeEnabled' => true,
        'explodeParameterName' => 'explode',
        'separatorParameterName' => 'separator',
    ]
)]
```
Assuming the database contains 2 users:
- A) Jon Doe
- B) Sara Doe

Searching for "Jon" =>  will return only "Jon Doe"
`/users.json?search=Jon`
Searching for "Jon doe" => will return both users. (space acting as OR ... like ...)
`/users.json?search=Jon%20doe`
Searching for "Jon,doe" => will return only "Jon Doe" (comma acting as AND ... like ...)
`/users.json?search=Jon%2Cdoe` or `/users.json?search[]=Jon&search[]=doe`

**Comment by metaclass-nl** [comment]
I found the following (mostly other issues etc. here):

- https://github.com/soyuka/esql
- https://github.com/rollerworks/search
- https://github.com/krakphp/aql
- https://www.odata.org/documentation/odata-version-2-0/uri-conventions/

And then there is the api of the existing ORM Filter classes. They can become composable by implementing [an extra interface](https://github.com/metaclass-nl/core/blob/query-expression-generator/src/Bridge/Doctrine/Orm/Filter/QueryExpressionGeneratorInterface.php), see [my comment on RFC Filter composition](https://github.com/api-platform/core/issues/2400#issuecomment-850821749)

For a Filter whose composition can be configured one could do something similar to this [WhereFilter (FilterLogic)](https://github.com/metaclass-nl/filter-bundle/tree/query-expression-generator), but as input use an array from the configuration specifying nested logic and the results of evaluating Symfony Expressions that have access to $context. Such an expression will throw an exception if a key from $context it accesses is not present. That exception should be caught resulting in the result of only that expression not getting added to the logic. Here is an example:
```php docblock
@ApiFilter(SearchFilter:class, properties={"sku": "iexact", "name": "iword_start", "description": "iword_start"})
@ApiFilter(DateFilter::class, properties={"published": DateFilter::EXCLUDE_NULL})
@ApiFilter(Compose:class, arguments={
	"composition": {
		"or": {
			"filter('sku', context.filters.q)",
			"filter('name', filters.q)",
			"filter('description', context.filters.q)",
			"logic.and(
				filter('published[after]', context.filters.from),
				filter('published[before]', context.filters.through)
			)"
		}
	},
	"apply": "andWhere"
})
```
